### PR TITLE
Don't return empty pages with next links

### PR DESCRIPTION
**Pull request by alexpaulzor**
@bcavagnolo MC-459

### CODE REVIEW

File: datalake_api/fetcher.py [hunk 1]
```diff
@@ -84,7 +84,8 @@ def _get_s3_key(self, file_id):
         try:
             return self.s3_bucket.Object(path).get()
         except BotoClienError as e:
-            if e.response['ResponseMetadata']['HTTPStatusCode'] == 404:
+            if 'ResponseMetadata' in e.response and \
+                    e.response['ResponseMetadata']['HTTPStatusCode'] == 404:
                 msg = 'No file with id {} exists'.format(file_id)
```
Comment (bcavagnolo) [inline datalake_api/fetcher.py]: Did you encounter a case where ResponseMetadata was not present? If so, maybe there's a boto library difference we should know more about. Specifically, if there's another way to get a 404, we should accommodate it.

File: datalake_api/querier.py [hunk 2]
```diff
@@ -210,20 +210,22 @@ def query_by_time(self, start, end, what, where=None, cursor=None):
...
+                    self._add_cursor_conditions(kwargs, cursor)
+
+                response = self._table.query(**kwargs)
+                results += self._exclude_outside(response['Items'], start, end)
+                # we _could_ deduplicate the results here to make more headroom
+                # for another bucket.
+                cursor = self._cursor_for_time_query(response, results, b)
+                if cursor is None:
+                    # no more results in the bucket
+                    break
```
Comment (bcavagnolo) [inline datalake_api/querier.py]: Let's see if we can make this more straightforward by just tuning the limit parameter properly. I'm not really sure why I had all of this `MAX_RESULTS/2` business when we have a limit parameter at our disposal. How about something like this:

```
for b in buckets:
    headroom = MAX_RESULTS - len(results)
    if headroom <= 0 or cursor is None:
        break
    kwargs = dict(
        bucket=b,
        what=what,
        start=start, 
        end=end,
        where=where,
        limit=headroom, cursor
    )
    response = self._query_time_bucket(**kwargs))
    results += response['Items']
    cursor = self._cursor_for_time_query(response, results, b)
```

_query_time_bucket would have its own inner loop that repeatedly queries the bucket if there's more data and more headroom. Unfortunately, I don't think we are going to be able to articulate the behavior of `_exclude_outside()` as a filter condition. So in the case where there are more than MAX_RESULTS of the desired `what`s in a time bucket that do not lie within the requested time window we have to do multiple queries in that bucket.

File: tests/test_archive_querier.py [hunk 3]
```diff
@@ -238,8 +238,8 @@ def get_multiple_pages(query_function, query_args):
         page = query_function(*query_args, cursor=cursor)
         page_len = len(page)
         assert page_len <= MAX_RESULTS
-        if len(results) == 0:
-            assert page.cursor is not None
```
Comment (bcavagnolo) [inline tests/test_archive_querier.py]: We actually want to leave this condition as-is. It checks that the first page of the multi-page request has a cursor. I think you should add another test condition that looks like this:

```
if page.cursor is not None:
    assert page_len > 0
```

This says that a page without results can only be a final page with no cursor.

File: datalake_api/fetcher.py (hunk 1 above)
Comment (alexpaulzor) [inline datalake_api/fetcher.py]: I did, but don't recall. It was while testing on my mac, outside of a virtualenv. I'll stop doing that and remove this change.

File: datalake_api/querier.py (hunk 2 above)
Comment (alexpaulzor) [inline datalake_api/querier.py]: Ok yeah, that was my original plan, but I remember suggesting it and getting shot down, though I don't recall specifically.

File: tests/test_archive_querier.py [hunk 4]
```diff
@@ -290,6 +290,21 @@ def test_paginate_many_records_single_time_bucket(table_maker, querier):
...
+    very_end = DatalakeRecord.TIME_BUCKET_SIZE_IN_MS - 1
+    for start in range(0, very_end, interval):
+        end = start + interval
+        records += create_test_records(start=start, end=end, what='foo')
+    table_maker(records)
+    results = get_multiple_pages(
+        querier.query_by_time,
+        [very_end - 2 * interval + 2, very_end, 'foo'])
+    evaluate_time_based_results(results, 2)
+
```
Comment (bcavagnolo) [inline tests/test_archive_querier.py]: I'm not convinced that this test reliably reproduced the condition we are concerned about. I pasted this test code into master (along with my modified test condition above) and it only seems to fail like 10% of the time.

File: tests/test_archive_querier.py [hunk 5]
```diff
@@ -231,15 +231,19 @@ def test_deduplicating_work_id_records(table_maker, querier):
     assert len(results) == 1
 
 
-def get_multiple_pages(query_function, query_args):
+def get_multiple_pages(query_function, query_args, allow_single_page=False):
```
Comment (bcavagnolo) [inline tests/test_archive_querier.py]: Can we drop the boolean allow_single_page=True argument? boolean arguments like this are usually a signal that you should really have two functions. And you don't seem to use the new argument anyway.

File: tests/test_archive_querier.py [hunk 6]
```diff
@@ -231,15 +231,19 @@ def test_deduplicating_work_id_records(table_maker, querier):
...
     results = []
     cursor = None
     while True:
         page = query_function(*query_args, cursor=cursor)
         page_len = len(page)
         assert page_len <= MAX_RESULTS
-        if len(results) == 0:
+        # Ensure the first page has a cursor or is not empty
+        if len(results) == 0 and not allow_single_page:
```
Comment (bcavagnolo) [inline tests/test_archive_querier.py]: ...and here I feel like the test condition should be:
```
if cursor is not None:
    assert len(results) > 0
```

File: tests/test_archive_querier.py [hunk 7]
```diff
@@ -290,6 +294,56 @@ def test_paginate_many_records_single_time_bucket(table_maker, querier):
...
+    very_end = DatalakeRecord.TIME_BUCKET_SIZE_IN_MS
+    for start in range(0, very_end, interval):
+        end = start + interval
+        records += create_test_records(start=start, end=end, what='foo')
+    table_maker(records)
+    results = get_multiple_pages(
+        querier.query_by_time,
+        [very_end - 3 * interval + 1,
+         very_end - interval - 1, 'foo'],
+        True)
```
Comment (bcavagnolo) [inline tests/test_archive_querier.py]: "The first 2 of the last 3 records?" Dizzying. Why can't we just target the last record? `start` should land just where you want it. Then this query becomes something like:

```
results = get_multiple_pages(querier.query_by_time, [start, start + 1, 'foo'])
```

I tried this. It does seem to reproduce the error (along with the test condition I proposed above).

File: tests/test_archive_querier.py [hunk 8]
```diff
@@ -290,6 +294,56 @@ def test_paginate_many_records_single_time_bucket(table_maker, querier):
...
+        start=DatalakeRecord.TIME_BUCKET_SIZE_IN_MS*11/4,
+        end=DatalakeRecord.TIME_BUCKET_SIZE_IN_MS*11/4+1, what='foo')
+
+    table_maker(records)
+    start = DatalakeRecord.TIME_BUCKET_SIZE_IN_MS * 3 / 4
+    end = DatalakeRecord.TIME_BUCKET_SIZE_IN_MS * 9 / 4
+    results = get_multiple_pages(
+        querier.query_by_time,
+        [start,
+         end, 'foo'],
```
Comment (bcavagnolo) [inline tests/test_archive_querier.py]: Nit: can you keep this short list argument on a single line?

**Comment by bcavagnolo** [comment]
Thanks @alexpaulzor!

**Pull request merged**